server/tracedb/connection/tester.go

```diff
@@ -0,0 +1,48 @@
+package connection
+
+import "context"
+
+type TestStep interface {
+	TestConnection(ctx context.Context) ConnectionTestStepResult
+}
+
+type TesterOption func(*Tester)
+
+type Tester struct {
+	connectivityTestStep   TestStep
+	authenticationTestStep TestStep
+	pollingTestStep        TestStep
+}
+
+func NewTester(opts ...TesterOption) Tester {
+	tester := Tester{}
+
+	for _, opt := range opts {
+		opt(&tester)
+	}
+
+	return tester
+}
+
+func (t *Tester) TestConnection(ctx context.Context) ConnectionTestResult {
+	connectivityTestResult := t.connectivityTestStep.TestConnection(ctx)
+	if connectivityTestResult.Error != nil {
+		return ConnectionTestResult{
+			ConnectivityTestResult: connectivityTestResult,
+		}
+	}
+
+	authTestResult := t.authenticationTestStep.TestConnection(ctx)
+	if authTestResult.Error != nil {
+		return ConnectionTestResult{
+			ConnectivityTestResult:   connectivityTestResult,
+			AuthenticationTestResult: authTestResult,
+		}
+	}
+
+	return ConnectionTestResult{
+		ConnectivityTestResult:   t.connectivityTestStep.TestConnection(ctx),
+		AuthenticationTestResult: t.authenticationTestStep.TestConnection(ctx),
+		TraceRetrievalTestResult: t.pollingTestStep.TestConnection(ctx),
+	}
```

Comment:
This code is running the tests twice, once to check the error and again for building the result. It can do both things at the same time:

```suggestion
func (t *Tester) TestConnection(ctx context.Context) res ConnectionTestResult {
	res.ConnectivityTestResult = t.connectivityTestStep.TestConnection(ctx)
	if res.ConnectivityTestResult.Error != nil {
		return 
	}

	res.AuthenticationTestResult = t.authenticationTestStep.TestConnection(ctx)
	if res.ConnectivityTestResult.Error != nil {
		return 
	}

	res.TraceRetrievalTestResult = t.pollingTestStep.TestConnection(ctx)

	return
}
```